### PR TITLE
NIFI-13724 Add additional metadata to Registry events

### PR DESCRIPTION
**Pull request by Dye357**
# Summary

[NIFI-13724](https://issues.apache.org/jira/browse/NIFI-13724)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X ] Pull Request based on current revision of the `main` branch
- [X ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ X] Build completed using `mvn clean install -P contrib-check`
  - [ X] JDK 21

### Licensing

- [ n/a] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ n/a] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ X] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java [hunk 1]
```diff
@@ -37,6 +37,10 @@ public static Event bucketCreated(final Bucket bucket) {
         return new StandardEvent.Builder()
                 .eventType(EventType.CREATE_BUCKET)
                 .addField(EventFieldName.BUCKET_ID, bucket.getIdentifier())
+                .addField(EventFieldName.BUCKET_NAME, bucket.getName())
+                .addField(EventFieldName.BUCKET_DESCRIPTION, bucket.getDescription())
```
Comment (bbende) [inline nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java:41]: The description of a bucket is optional and can be null. Currently the event class doesn't accept null values so I think we'll have to make this conditional to only set the description if not null, or maybe set empty string in the event if null.

Running the integration tests for registry produces a failing test with:
```
[qtp1243712710-92] ERROR org.apache.nifi.registry.web.mapper.ThrowableMapper - An unexpected error. Returning Internal Server Error response.
java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:233)
	at org.apache.nifi.registry.event.StandardEventField.<init>(StandardEventField.java:35)
	at org.apache.nifi.registry.event.StandardEvent$Builder.addField(StandardEvent.java:96)
	at org.apache.nifi.registry.event.EventFactory.bucketCreated(EventFactory.java:41)
	at org.apache.nifi.registry.web.api.BucketResource.createBucket(BucketResource.java:94)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at **java.base/java.lang.reflect.Method.invoke(Method.java:580)**
```
From:
```
[ERROR] Errors:
[ERROR]   DBFlowStorageIT.testAll:94 » NiFiRegistry Error creating bucket: An unexpected error has occurred. Please check the logs for additional details.
```

`mvn clean install -Pintegration-tests` from `nifi/nifi-registry`

File: nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java [hunk 2]
```diff
@@ -45,6 +49,10 @@ public static Event bucketUpdated(final Bucket bucket) {
         return new StandardEvent.Builder()
                 .eventType(EventType.UPDATE_BUCKET)
                 .addField(EventFieldName.BUCKET_ID, bucket.getIdentifier())
+                .addField(EventFieldName.BUCKET_NAME, bucket.getName())
+                .addField(EventFieldName.BUCKET_DESCRIPTION, bucket.getDescription())
```
Comment (bbende) [inline nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java:53]: Handle null

File: nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java [hunk 3]
```diff
@@ -53,6 +61,10 @@ public static Event bucketDeleted(final Bucket bucket) {
         return new StandardEvent.Builder()
                 .eventType(EventType.DELETE_BUCKET)
                 .addField(EventFieldName.BUCKET_ID, bucket.getIdentifier())
+                .addField(EventFieldName.BUCKET_NAME, bucket.getName())
+                .addField(EventFieldName.BUCKET_DESCRIPTION, bucket.getDescription())
```
Comment (bbende) [inline nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java:65]: Handle null

File: nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java [hunk 4]
```diff
@@ -61,7 +73,13 @@ public static Event flowCreated(final VersionedFlow versionedFlow) {
         return new StandardEvent.Builder()
                 .eventType(EventType.CREATE_FLOW)
                 .addField(EventFieldName.BUCKET_ID, versionedFlow.getBucketIdentifier())
+                .addField(EventFieldName.BUCKET_NAME, versionedFlow.getBucketName())
                 .addField(EventFieldName.FLOW_ID, versionedFlow.getIdentifier())
+                .addField(EventFieldName.FLOW_NAME, versionedFlow.getName())
+                .addField(EventFieldName.FLOW_DESCRIPTION, versionedFlow.getDescription())
```
Comment (bbende) [inline nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java:79]: Handle null

File: nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java [hunk 5]
```diff
@@ -70,7 +88,12 @@ public static Event flowUpdated(final VersionedFlow versionedFlow) {
         return new StandardEvent.Builder()
                 .eventType(EventType.UPDATE_FLOW)
                 .addField(EventFieldName.BUCKET_ID, versionedFlow.getBucketIdentifier())
+                .addField(EventFieldName.BUCKET_NAME, versionedFlow.getBucketName())
                 .addField(EventFieldName.FLOW_ID, versionedFlow.getIdentifier())
+                .addField(EventFieldName.FLOW_NAME, versionedFlow.getName())
+                .addField(EventFieldName.FLOW_DESCRIPTION, versionedFlow.getDescription())
```
Comment (bbende) [inline nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java:94]: Handle null

File: nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java [hunk 6]
```diff
@@ -79,7 +102,12 @@ public static Event flowDeleted(final VersionedFlow versionedFlow) {
         return new StandardEvent.Builder()
                 .eventType(EventType.DELETE_FLOW)
                 .addField(EventFieldName.BUCKET_ID, versionedFlow.getBucketIdentifier())
+                .addField(EventFieldName.BUCKET_NAME, versionedFlow.getBucketName())
                 .addField(EventFieldName.FLOW_ID, versionedFlow.getIdentifier())
+                .addField(EventFieldName.FLOW_NAME, versionedFlow.getName())
+                .addField(EventFieldName.FLOW_DESCRIPTION, versionedFlow.getDescription())
```
Comment (bbende) [inline nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java:108]: Handle null

**Review by bbende — COMMENTED** [review]
General idea seems good, just a few comments about handling optional fields, and a question about whether `FLOW_VERSION` is really needed to be added.

File: nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java [hunk 7]
```diff
@@ -61,7 +73,13 @@ public static Event flowCreated(final VersionedFlow versionedFlow) {
         return new StandardEvent.Builder()
                 .eventType(EventType.CREATE_FLOW)
                 .addField(EventFieldName.BUCKET_ID, versionedFlow.getBucketIdentifier())
+                .addField(EventFieldName.BUCKET_NAME, versionedFlow.getBucketName())
                 .addField(EventFieldName.FLOW_ID, versionedFlow.getIdentifier())
+                .addField(EventFieldName.FLOW_NAME, versionedFlow.getName())
+                .addField(EventFieldName.FLOW_DESCRIPTION, versionedFlow.getDescription())
+                .addField(EventFieldName.CREATED_TIMESTAMP, String.valueOf(versionedFlow.getCreatedTimestamp()))
+                .addField(EventFieldName.MODIFIED_TIMESTAMP, String.valueOf(versionedFlow.getModifiedTimestamp()))
+                .addField(EventFieldName.FLOW_VERSION, String.valueOf(versionedFlow.getRevision().getVersion()))
```
[the remark this answers is not in the document]
Comment (Dye357) [reply nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java]: I believe you are correct - I'm making the requested changes and will re-test our use-case for this field.

[the remark this answers is not in the document]
Comment (Dye357) [reply nifi-registry/nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/event/EventFactory.java]: Integration tests passed for me locally - I also addressed the null cases and added additional test-cases.

**Comment by Dye357** [comment]
Just pushed a new commit to fix a whitespace issue.

**Review by bbende — APPROVED** [review]
Latest updates look good, thanks for following up with the changes

**Comment by Dye357** [comment]
Thanks @bbende! What would I need to do to have this PR also dropped onto the 1.x support branch. Ideally for the 1.28 release?

**Comment by bbende** [comment]
@Dye357 At this point it this point it mostly requires two separate PRs, one to `main` and then one to `support/nifi-1.x`. For most NiFi specific changes, the cherry pick doesn't apply cleanly because everything has changed, but since this only registry, hopefully you can just create a local branch from `support/nifi-1.x`, cherry pick the commit, push the branch to your fork and create another PR against `support/nifi-1.x`.

**Pull request merged**